File: EnvelopeASP/Controllers/TransactionController.cs
using EnvelopeASP.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Rendering;
using Microsoft.Extensions.Primitives;

namespace EnvelopeASP.Controllers
{
    [Authorize]
    [HTMXDefaultLayoutFilter]
    public class TransactionController : Controller
    {
        public async Task<IActionResult> Index(int? id, uint page)
        {
            if(id.HasValue == false)
            {
                return Redirect("/Envelope");
            }
            var uid = Utils.GetUserIDFromClaims(HttpContext.User);

            page = Math.Max(1, page);

            var transactions = await Procedures.Sel_Transactions(uid, Convert.ToUInt16(Math.Abs(id.Value)), Utils.DEFAULT_PAGINATION_SIZE, page);
            return View((id.Value,transactions, page));
        }

        public async Task<IActionResult> Add(int? id, [FromForm]Transaction model)
        {
            ViewData["Operation"] = "Add";
            if(id.HasValue == false)
            {
                return RedirectToAction("Index");
            }
            if(Request.Method == System.Net.WebRequestMethods.Http.Post)
            {
                if (ModelState.IsValid == false)
                {
                    return View(model);
                }
                var uID = Utils.GetUserIDFromClaims(HttpContext.User);
                var eID = Convert.ToUInt16(Math.Abs(id.Value));

                //todo, dollar to cent conversion
                await Procedures.Ins_EnvelopeTransaction(uID, eID, Utils.DoubleMoneyToCents(model.Amount), model.Date, model.Note);

                var hasAddAgain = Request.Form.TryGetValue("AddAgain", out StringValues addAgainValue);
                if (hasAddAgain && addAgainValue.Equals("on"))
                {
                    model = new Transaction() { Date = model.Date };
                } else
                {
                    if (Utils.RequestIsHTMX(HttpContext))
                    {
                        Response.Headers.Add("HX-Refresh", "true");
                        return View(model);
                    }
                    
                    return RedirectToAction("Index", new { id });
                }
            }

            model ??= new Transaction();
            return View(model);
        }

        public async Task<IActionResult> Transfer(int? id, [FromForm] TransferModel model)
        {
            if (id.HasValue == false)
            {
                return RedirectToAction("Index");
            }
            if (Request.Method == System.Net.WebRequestMethods.Http.Post)
            {
                if (ModelState.IsValid == false)
                {
                    return View(model);
                }
                if(model.DestinationNumber.HasValue == false)
                {
                    return View(model);
                }
                var uID = Utils.GetUserIDFromClaims(HttpContext.User);
                var eID = Convert.ToUInt16(Math.Abs(id.Value));
                var dID = Convert.ToUInt16(Math.Abs(model.DestinationNumber.Value));

                await Procedures.Transfer(uID, eID, dID, Utils.DoubleMoneyToCents(model.Amount));
                return RedirectToAction("Index", new { id });
            }

            model ??= new TransferModel();
            if(model.Envelopes.Count == 0)
            {
                var envelopes = await Procedures.Sel_Envelope_Summary(Utils.GetUserIDFromClaims(HttpContext.User));
                model.Envelopes = envelopes.Where(f => f.Number != id).Select(f => new SelectListItem(f.Name + " - " + f.Amount, f.Number.ToString())).ToList();
            }
            return View(model);
        }

        [Route("/Transaction/Update/{id}/{tid}")]
        public async Task<IActionResult> Update(ushort id, uint tid, [FromForm] Transaction model)
        {
            ViewData["Operation"] = "Update";
            var uID = Utils.GetUserIDFromClaims(HttpContext.User);
            if (Request.Method == System.Net.WebRequestMethods.Http.Post)
            {
                if (ModelState.IsValid == false)
                {
                    return View("Add", model);
                }

                await Procedures.Upd_EnvelopeTransaction(uID, id, tid, Utils.DoubleMoneyToCents(model.Amount), model.Date, model.Note);
                if (Utils.RequestIsHTMX(HttpContext))
                {
                    Response.Headers.Add("HX-Refresh", "true");
                    return View("Add", model);
                }
                return Redirect("/Transaction/Index/" + id);
            }
            if(model.TransactionNumber == 0)
            {

                var nmodel = await Procedures.Sel_Transaction(uID, id, tid);
                if(nmodel == null)
                {
                    return Redirect("/Transaction/Index/" + id);
                }
                model = nmodel;
            }
            
            return View("Add", model);
        }

        [Route("/Transaction/Delete/{eid}/{tid}")]
        public async Task<IActionResult> Delete(ushort eid, uint tid, [FromForm] Transaction model)
        {
            var uID = Utils.GetUserIDFromClaims(HttpContext.User);
            if (Request.Method == System.Net.WebRequestMethods.Http.Post)
            {
                var yesno = Request.Form.FirstOrDefault(f => f.Key == "yesno");
                var val = yesno.Value.ToString();
                if (string.IsNullOrEmpty(val))
                {
                    return View(model);
                }
                if(val.Equals("Yes", StringComparison.OrdinalIgnoreCase))
                {
                    await Procedures.Del_EnvelopeTransaction(uID, eid, tid);
                }
                if (Utils.RequestIsHTMX(HttpContext))
                {
                    Response.Headers.Add("HX-Refresh", "true");
                    return View(model);
                }
                return Redirect("/Transaction/Index/" + eid);
            }
            if (model.TransactionNumber == 0)
            {

                var nmodel = await Procedures.Sel_Transaction(uID, eid, tid);
                if (nmodel == null)
                {
                    return RedirectToAction("Index", new { id = ((int)eid) });
                }
                model = nmodel;
            }

            return View(model);
        }
    }
}

File: EnvelopeWeb/Controllers/TransactionController.fs
namespace EnvelopeWeb.Controllers

open Microsoft.AspNetCore.Mvc
open Microsoft.Extensions.Logging
open System.Threading.Tasks
open Models

[<Microsoft.AspNetCore.Authorization.AuthorizeAttribute>]
[<EnvelopeWeb.HTMXDefaultLayoutFilterAttribute>]
type TransactionController (logger : ILogger<TransactionController>, dbConnectionGetter : DbConnection.IDbConnectionGetter) =
    inherit Controller()

    member this.Index (id: System.Nullable<uint16>) (page: uint32) =
        async {
            if id.HasValue = false then
                return this.Redirect("/Envelope") :> IActionResult
            else
                let page = System.Math.Max(1u, page)
                let uid = Utils.getUserIDFromClaims this.User
                let! transactions = Procedures.Sel_Transactions dbConnectionGetter uid id.Value Utils.DefaultPaginationSize page

                return this.View((id.Value, transactions, page) |> System.TupleExtensions.ToValueTuple) :> IActionResult
        }
        |> Async.StartAsTask
        
    member this.Add (id:System.Nullable<uint16>) ([<FromForm>]model: Models.Transaction) : Task<IActionResult> =
        this.ViewData["Operation"] <- "Add"
        async {
            if id.HasValue = false then
                return this.RedirectToAction("Index") :> IActionResult
            else if this.Request.Method = System.Net.WebRequestMethods.Http.Post && this.ModelState.IsValid then
                let uid = Utils.getUserIDFromClaims this.User
                
                let dollarAsCents = Utils.doubleMoneyToCents model.Amount

                let! r = Procedures.Ins_EnvelopeTransaction dbConnectionGetter uid id.Value dollarAsCents model.Date model.Note

                let hasAddAgain, hasAddAgainValue = this.Request.Form.TryGetValue "AddAgain"

                if hasAddAgain && hasAddAgainValue.Equals("on") then 
                    let model = { Amount = 0.0; Note = System.String.Empty; Date = model.Date; TransactionNumber = 0u }
                    return this.View(model) :> IActionResult
                else if Utils.requestIsHTMX this.HttpContext then
                    this.Response.Headers.Add("HX-Refresh", "true")
                    return this.View(model)
                else
                    return this.RedirectToAction("Index", {| id = id |}) :> IActionResult
                        
            else
                return this.View(model) :> IActionResult
                
        }
        |> Async.StartAsTask

    member this.Transfer (id: uint16) ([<FromForm>]model:Models.Transfer) =
        async {

            if this.Request.Method = System.Net.WebRequestMethods.Http.Post && this.ModelState.IsValid && model.DestinationNumber.HasValue then
                let uid = Utils.getUserIDFromClaims this.User
                let amountAsCents = Utils.doubleMoneyToCents model.Amount
                let! r = Procedures.Transfer dbConnectionGetter uid id model.DestinationNumber.Value amountAsCents

                return this.RedirectToAction("Index", {| id = id |}) :> IActionResult
            else
                let uid = Utils.getUserIDFromClaims this.User
                let! env = Procedures.Sel_Envelope_Summary dbConnectionGetter uid
                let env = env |> List.filter (fun f -> f.Number <> id) |> List.map (fun f -> new Microsoft.AspNetCore.Mvc.Rendering.SelectListItem(f.Name + " - " + f.Amount.ToString(), f.Number.ToString()))
                let model = { DestinationNumber = new System.Nullable<uint16>(); Amount = 0.0; Envelopes = env }
                return this.View(model) :> IActionResult
                
        }
        |> Async.StartAsTask


    [<Route("/Transaction/Update/{id}/{tid}")>]
    member this.Update (id:uint16) (tid:uint) ([<FromForm>]model:Models.Transaction) =
        this.ViewData["Operation"] <- "Update"
        async {
            if this.Request.Method = System.Net.WebRequestMethods.Http.Post && this.ModelState.IsValid then
                let uid = Utils.getUserIDFromClaims this.User
                
                let dollarAsCents = Utils.doubleMoneyToCents model.Amount

                let! r = Procedures.Upd_EnvelopeTransaction dbConnectionGetter uid id tid dollarAsCents model.Date model.Note

                if Utils.requestIsHTMX this.HttpContext then
                    this.Response.Headers.Add("HX-Refresh", "true")
                    return this.View("Add", model) :> IActionResult
                else
                    return this.Redirect("/Transaction/Index/" + id.ToString()) :> IActionResult
            
            else if model.TransactionNumber = 0u then
                let uid = Utils.getUserIDFromClaims this.User
                let! model = Procedures.Sel_Transaction dbConnectionGetter uid id tid
                match model with
                | None -> return this.Redirect("/Transaction/Index/" + id.ToString()) :> IActionResult
                | Some m -> return this.View("Add", m) :> IActionResult
            else
                return this.View("Add", model) :> IActionResult
        }
        |> Async.StartAsTask

    [<Route("/Transaction/Delete/{eid}/{tid}")>]
    member this.Delete (eid:uint16) (tid:uint) ([<FromForm>]model:Models.Transaction) =
        let uid = Utils.getUserIDFromClaims this.User
        async {
            if this.Request.Method = System.Net.WebRequestMethods.Http.Post then
                let hasYesNo, yesno = this.Request.Form.TryGetValue "yesno"
                if hasYesNo = false || System.String.IsNullOrWhiteSpace(yesno.Item 0) then
                    return this.View(model) :> IActionResult
                else
                    if "Yes".Equals(yesno.Item 0, System.StringComparison.OrdinalIgnoreCase) then
                        let! r = Procedures.Del_EnvelopeTransaction dbConnectionGetter uid eid tid
                        r |> ignore
                    
                    if Utils.requestIsHTMX this.HttpContext then
                        this.Response.Headers.Add("HX-Refresh", "true")
                        return this.View(model) :> IActionResult
                    else
                        return this.Redirect("/Transaction/Index/" + eid.ToString()) :> IActionResult

            else
                let! dbModel = Procedures.Sel_Transaction dbConnectionGetter uid eid tid
                match dbModel with
                | None -> return this.RedirectToAction("Index", {| id = eid |}) :> IActionResult
                | Some m -> return this.View(m) :> IActionResult

                

        } |> Async.StartAsTask